### PR TITLE
plutip/servers: initial impl startCtlServer

### LINKED ISSUE #562: Introduce ctl-server to plutip testing environment
Part of #426

We want to configure `ctl-server` such that it uses the `cardano-node` instance which is launched by `plutip-server`.

 The Ask:

Modify `Plutip/Server.purs`, adding code to start up ctl-server from PATH with the appropriate parameters. Provide these parameters to the `ContractConfig` (see `mkClusterContractCfg`).

Relevant branches:

https://github.com/mlabs-haskell/plutip/pull/79 on plutip side
https://github.com/Plutonomicon/cardano-transaction-lib/pull/470 on CTL side <- make a PR to this branch

### PR DESCRIPTION
**Pull request by marijanp**
Closes #562

### Pre-review checklist

- [x] All code has been formatted using our config (`make format` for Purescript, `nixpkgs-fmt` for Nix)
- [x] All Purescript imports are explicit, including constructors
- [x] **All** existing examples have been run locally against a fully-synced testnet node
- [x] The integration and unit tests have been run (`npm run test`) locally
- [ ] The changelog has been updated under the `## Unreleased` header, using the appropriate sub-headings (`### Added`, `### Removed`, `### Fixed`)

### CODE REVIEW

File: src/Plutip/Server.purs [hunk 1]
```diff
@@ -12,35 +21,38 @@ import Contract.Monad (Contract, ContractConfig(ContractConfig), runContract)
 import Control.Monad.Error.Class (withResource)
 import Control.Monad.Reader (ReaderT(ReaderT), runReaderT)
 import Data.Bifunctor (lmap)
-import Data.Either (Either(Left), either)
+import Data.Either (Either(..), either, fromRight)
```
Comment (klntsky) [inline src/Plutip/Server.purs:24]: We use explicit imports everywhere, even for types like Either, please replace `..` with actually needed constructors.

File: src/Plutip/Server.purs [hunk 2]
```diff
@@ -216,8 +230,49 @@ type CtlServerConfig =
...
+
+-- eihter' :: (a -> c) -> Either a b -> (b -> c) -> c
+
+getNetworkInfoFromNodeConfig :: FilePath -> Effect NetworkInfo
+getNetworkInfoFromNodeConfig nodeConfigPath = do
+  nodeConfigJson <- either (\e -> throw $ show e) pure =<<
+    parseJsonStringToAeson <$> readTextFile Encoding.UTF8 nodeConfigPath
+  let
+    byronGenesisFile = getField
+      (unsafePartial $ fromJust $ toObject nodeConfigJson)
```
Comment (klntsky) [inline src/Plutip/Server.purs:261]: We should not use `unsafePartial`. Please use `fromMaybe (throw "some error message")` - all possible failure points must be explicitly handled. I think you can also parse everything is one go in monadic context. Also please use `.:` (a synonym for `getField`)

File: src/Plutip/Server.purs [hunk 3]
```diff
@@ -216,8 +230,49 @@ type CtlServerConfig =
...
+      "ByronGenesisFile"
+    requiresNetworkMagic = getField
+      (unsafePartial $ fromJust $ toObject nodeConfigJson)
+      "RequiresNetworkMagic"
+  if requiresNetworkMagic == Right "RequiresNoMagic" then pure Mainnet
+  else do
+    byronGenesisJson <- either (\e -> throw $ show e) pure =<<
```
Comment (klntsky) [inline src/Plutip/Server.purs:268]: `throw <<< show`

File: src/Plutip/Server.purs [hunk 4]
```diff
@@ -216,8 +230,49 @@ type CtlServerConfig =
...
+      parseJsonStringToAeson <$> readTextFile Encoding.UTF8
+        (fromRight "" byronGenesisFile)
```
Comment (klntsky) [inline src/Plutip/Server.purs:270]: We should throw if there is no byron genesis file instead of using `""`

File: src/Plutip/Server.purs [hunk 5]
```diff
@@ -216,8 +230,49 @@ type CtlServerConfig =
...
+  where
+  getNetworkInfoArgs :: NetworkInfo -> Array String
+  getNetworkInfoArgs Mainnet = []
+  getNetworkInfoArgs (Testnet networkId) =
+    [ "--network-id", Int.toStringAs Int.decimal networkId ]
+
+stopCtlServer :: ChildProcess -> Aff Unit
+stopCtlServer = liftEffect <<< kill SIGINT
+
+-- eihter' :: (a -> c) -> Either a b -> (b -> c) -> c
```
Comment (klntsky) [inline src/Plutip/Server.purs:253]: Please remove this

File: src/Plutip/Server.purs [hunk 6]
```diff
@@ -216,8 +230,49 @@ type CtlServerConfig =
...
+
+getNetworkInfoFromNodeConfig :: FilePath -> Effect NetworkInfo
+getNetworkInfoFromNodeConfig nodeConfigPath = do
+  nodeConfigJson <- either (\e -> throw $ show e) pure =<<
```
Comment (klntsky) [inline src/Plutip/Server.purs:257]: `throw <<< show`

**Review by klntsky — CHANGES_REQUESTED** [review]
Please also provide the appropriate server config to contract config constructor function instead of the default one: https://github.com/Plutonomicon/cardano-transaction-lib/blob/73330fc2e8efbb0059e0efea6efcd9d5320b5f93/src/Plutip/Server.purs#L206

File: src/Plutip/Server.purs (hunk 6 above)
Comment (marijanp) [reply src/Plutip/Server.purs:257]: I was looking for that :+1:

File: src/Plutip/Server.purs (hunk 2 above)
Comment (marijanp) [reply src/Plutip/Server.purs:261]: removed

File: src/Plutip/Server.purs [hunk 7]
```diff
@@ -216,8 +230,44 @@ type CtlServerConfig =
   , networkId :: NetworkInfo
   }
 
-runCtlServer :: CtlServerConfig -> Aff Unit
-runCtlServer = undefined
+startCtlServer :: PlutipConfig -> ClusterStartupParameters -> Aff ChildProcess
+startCtlServer cfg params = do
+  networkInfo <- liftEffect $ getNetworkInfoFromNodeConfig ""
```
Comment (klntsky) [inline src/Plutip/Server.purs:235]: Why is the file name empty?

Comment (marijanp) [reply src/Plutip/Server.purs:235]: Oh good catch

File: src/Plutip/Server.purs [hunk 8]
```diff
@@ -216,8 +233,44 @@ type CtlServerConfig =
...
+
+getNetworkInfoFromNodeConfig :: FilePath -> Effect NetworkInfo
+getNetworkInfoFromNodeConfig nodeConfigPath = do
+  nodeConfigYaml <- readTextFile Encoding.UTF8 nodeConfigPath
+  nodeConfigJson <- fromRightEff $ runExcept $ parseYAMLToJson nodeConfigYaml
+  nodeConfigAeson <- fromJustEff "Not a json object" $ toObject $ jsonToAeson
+    nodeConfigJson
+  byronGenesisFile <- fromRightEff $ getField nodeConfigAeson
+    "ByronGenesisFile"
+  let requiresNetworkMagic = getField nodeConfigAeson "RequiresNetworkMagic"
```
Comment (klntsky) [inline src/Plutip/Server.purs:264]: This variable can be inlined

**Review by klntsky — CHANGES_REQUESTED** [review]
Thank you, this is huge!

Could you please also replace localhost with 0.0.0.0 in Test.Plutip.config

File: src/Plutip/Server.purs [hunk 9]
```diff
@@ -189,21 +206,21 @@ mkClusterContractCfg plutipCfg _clusterParams = do
...
+        , host = plutipCfg.ogmiosDatumCacheConfig.host
         }
   usedTxOuts <- newUsedTxOuts
   pure $ ContractConfig
     { ogmiosWs
     , datumCacheWs
     , wallet: Nothing -- TODO: use KeyWallet when it's ready
     , usedTxOuts
-    , serverConfig: QueryM.defaultServerConfig
+    , serverConfig: plutipCfg.ctlServerConfig
```
Comment (marijanp) [inline src/Plutip/Server.purs:223]: @klntsky is that ok?

**Review by klntsky — APPROVED** [review]

**Pull request merged**